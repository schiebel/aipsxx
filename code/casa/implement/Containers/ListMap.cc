//# ListMap.cc: Map with list ordering/operations
//# Copyright (C) 1993,1994,1995,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: ListMap.cc,v 19.5 2004/12/01 16:42:17 ddebonis Exp $

#include <casa/Containers/ListMap.h>


namespace casa { //# NAMESPACE CASA - BEGIN

template<class key, class value> value *ListMapRep<key,value>::isDefined(const key &kv) {
  if (!list.atEnd() && list.getRight().x() == kv) {
    value &ret = list.getRight().y();
    return(&ret);
  }
  list.toStart();
  while (list.atEnd() != True){
    if ((list.getRight()).x() == kv) {
//
//      Had to change this because of a **SUN BUG**
//
//    return(&(list.getRight()).y());
//
      OrderedPair<key,value> &opTmp = list.getRight();
      return(&opTmp.y());
    }
    ++list;
  }
  return(0);
}

template<class key, class value> const value *ListMapRep<key,value>::isDefined(const key &kv) const {
  if (!list.atEnd() && list.getRight().x() == kv) {
    const value &ret = list.getRight().y();
    return(&ret);
  }
  ConstListIter<OrderedPair<key,value> > listp = list;
  listp.toStart();
  while (listp.atEnd() != True){
    if ((listp.getRight()).x() == kv) {
//
//      Had to change this because of a **SUN BUG**
//
//    return(&(list.getRight()).y());
//
      const OrderedPair<key,value> &opTmp = listp.getRight();
      return(&opTmp.y());
    }
    ++listp;
  }
  return(0);
}

template<class key, class value> uInt ListMapRep<key,value>::ndefined() const {
  return(list.len());
}

template<class key, class value> void ListMapRep<key,value>::remove(const key &kv) {
  if (!list.atEnd() && list.getRight().x() == kv){
    list.removeRight();
    return;
  }
  list.toStart();
  while (list.atEnd() != True) {
    if (list.getRight().x() == kv){
      list.removeRight();
      return;
    }
    ++list;
  }
}

template<class key, class value> value &ListMapRep<key,value>::define(const key &k, const value &v) {
  list.toStart();
  while (list.atEnd() != True) {
    if (list.getRight().x() == k){
      list.removeRight();
      break;
    }
    ++list;
  }
  if (deforder == Append)
    list.toEnd();
  else
    list.toStart();
  list.addRight(OrderedPair<key,value>(k,v));
  return(list.getRight().y());
}
    
template<class key, class value> void ListMapRep<key,value>::clear() {
  list.toStart();
  while (list.atEnd() != True){
    list.removeRight();
  }
}

template<class key, class value>
MapIterRep<key,value> *ListMapRep<key,value>::getRep(Map<key,value> *container) const {
  ListMapIterRep<key,value> *ret = new ListMapIterRep<key,value>((ListMap<key,value> *)container);
  return(ret);
}

template<class key, class value>
MapRep<key,value> *ListMapRep<key,value>::Clone() const {
  ListMapRep<key,value> *ret = new ListMapRep<key,value>(this->DefaultVal,deforder);
  ConstListIter<OrderedPair<key,value> > listp = list;

  ret->list.toStart();
  for(listp.toStart();listp.atEnd() != True; listp++) {
    ret->list.addRight(listp.getRight());
    ret->list++;
  }
  return ret;
}

template<class key, class value> 
void ListMapIterRep<key,value>::operator++() {
  if (listp.atEnd() == True)
    throw(AipsError("ListMap::operator++ - at end"));
  ++listp;
}

template<class key, class value>
void ListMapIterRep<key,value>::operator++(int) {
  if (listp.atEnd() == True)
    throw(AipsError("ListMap::operator++ - at end"));
  ++listp;
}

} //# NAMESPACE CASA - END

