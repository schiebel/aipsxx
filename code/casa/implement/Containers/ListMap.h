//# ListMap.h: Map with list ordering/operations
//# Copyright (C) 1993,1994,1995,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: ListMap.h,v 19.8 2005/06/18 21:19:14 ddebonis Exp $

#ifndef CASA_LISTMAP_H
#define CASA_LISTMAP_H

#include <casa/aips.h>
#include <casa/Containers/Map.h>
#include <casa/Containers/OrderedPair.h>
#include <casa/Containers/List.h>

namespace casa { //#Begin casa namespace

template<class key, class value> class ListMap;
template<class key, class value> class ListMapIterRep;

// <summary>Representation of a ListMap class</summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>

template<class key, class value> class ListMapRep : public MapRep<key,value> {
friend class ListMap<key,value>;
public:

  enum DefineOrder {Append, Prepend, Undefined};

  //
  //  Creates a map with a particular default value
  //
  //+grp
  ListMapRep(const value &dflt)
    : MapRep<key,value>(dflt),
      list(new List<OrderedPair<key,value> >,True), 
      deforder(Undefined)
    { }

  ListMapRep(const value &dflt, DefineOrder InOrder)
    : MapRep<key,value>(dflt),
      list(new List<OrderedPair<key,value> >,True), 
      deforder(InOrder)
    { }
  //-grp

  //
  // These functions check to see if a mapping is defined between
  // the specified key and some value. If one is, a pointer to
  // the value is returned, otherwise 0 is returned.
  //
  //+grp
  value *isDefined(const key &);
  const value *isDefined(const key &) const;
  //-grp

  //
  // Returns the number of user defined mappings
  //
  uInt ndefined() const;

  //
  //  Defines a mapping (ie. create a key value mapping)
  //
  value &define(const key &k, const value &v);

  //
  //  Undefines a mapping (ie. remove a key value mapping).
  //
  void remove(const key &);

  //
  // Clear all of the mappings.
  //
  void clear();

  MapIterRep<key,value> *getRep(Map<key,value> *) const;

  MapRep<key,value> *Clone() const;

  //
  // Set/Get the insertion order
  //
  //+grp
  DefineOrder getOrder() const { return deforder;}
  void setOrder(DefineOrder OR) { deforder = OR;}
  //-grp

  enum {ListMapRepVersion = 1};

protected:
  //
  //  The list used to store the key/value maps.
  //
  ListIter<OrderedPair<key,value> > list;

  DefineOrder deforder;

};

//
// <category lib=aips sect="Containers">
// <summary>Map with list ordering/operations</summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
//  This is one possible implementation of the Map class. It uses
//  a linked list to implement an associative array. Other likely 
//  implementations are hashed, and binary tree. It can be used
//  as follows:
//  <code>
//      ListMap<int,int> map(0);
//      map.define(3,4);
//      map.define(5,6);
//      map.define(7,8);
//      cout << map(3) << endl;
//      cout << map(7) << endl;
//      map(7) = 78;
//  </code>
//
template<class key, class value> class ListMap : public Map<key,value> {
friend class ListMapIterRep<key,value>;
protected:

  ListIter<OrderedPair<key,value> > &list() {
    return ((ListMapRep<key,value> *)(this->Rep))->list;
  }

public:

  enum DefineOrder {Append, Prepend, Undefined};

  //
  //  Creates a map with a particular default value
  //
  //+grp
  explicit ListMap(const value &dflt) : Map<key,value>(new ListMapRep<key,value>(dflt)) { }

  ListMap(const value &dflt, DefineOrder InOrder) : 
                          Map<key,value>(new ListMapRep<key,value>(dflt,
							InOrder == Append ? ListMapRep<key,value>::Append :
							InOrder == Prepend ? ListMapRep<key,value>::Prepend :
							ListMapRep<key,value>::Undefined)) { }
  //-grp

  //
  // Creates a ListMap from an existing one (copy semantics).
  //
  ListMap(const ListMap<key,value> &other) : Map<key,value>(other.Rep->Clone()) { }

  //
  // Assigns one list map to another (copy semantics).
  //
  ListMap<key,value> &operator=(const ListMap<key,value>&other) {
    SetRep(other.Rep->Clone());
    return *this;
  }

  //
  // Set/Get the insertion order
  //
  //+grp
  DefineOrder getOrder() const { 
    int v = ((ListMapRep<key,value> *)(this->Rep))->getOrder();
    return v == ListMapRep<key,value>::Append ? Append :
           v == ListMapRep<key,value>::Prepend ? Prepend :
           Undefined;
  }
  void setOrder(DefineOrder OR) { 
    ((ListMapRep<key,value> *)(this->Rep))->setOrder(OR == Append ? ListMapRep<key,value>::Append :
		  OR == Prepend ? ListMapRep<key,value>::Prepend :
		  ListMapRep<key,value>::Undefined);
  }
  //-grp

  enum {ListMapVersion = 1};

};

//
// <category lib=aips sect="Containers">
// <summary>ListMap iterator "letter"</summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// This is the "letter" which when paired (Const)MapIter "envelope"
// allows traversal of "ListMap"s.
//
template<class key, class value> class ListMapIterRep : virtual public MapIterRep<key,value> {
protected:

  //  The list used to store the key/value maps.
  ListIter<OrderedPair<key,value> > listp;

public:

  //
  // Checks to see if the iterator is in a valid state.
  //
  Bool isValid() const {return(listp.isValid());}

  //
  // Checks to see if the iterator is at one of the
  // map extremes, "atEnd()" or "atStart()".
  //
  //+grp
  Bool atEnd() const {return(listp.atEnd());}
  Bool atStart() const {return(listp.atStart());}
  //-grp

  //
  // Move the iterator to the beginning of the Map.
  //
  void toStart() {listp.toStart();}

  //
  // Advance the iterator to the next key.
  //
  //+grp
  void operator++();
  void operator++(int);
  //-grp

  //
  // Retrieve the key at the current iterator position.
  //
  const key &getKey() const {
    if (listp.atEnd() == True)
      throw(AipsError("ListMap::getKey - at end"));
    return(((ConstListIter<OrderedPair<key,value> >) listp).getRight().x());
  }

  //
  // Retrieve the value at the current iterator position.
  //
  //+grp
  const value &getVal() const {
    if (listp.atEnd() == True)
      throw(AipsError("ListMap::getVal - at end"));
    return(((ConstListIter<OrderedPair<key,value> >) listp).getRight().y());
  }
  value &getVal() {
    if (listp.atEnd() == True)
      throw(AipsError("ListMap::getVal - at end"));
    return(listp.getRight().y());
  }
  //-grp

  MapIterRep<key,value> *Clone() {
    ListMapIterRep<key,value> *ret = new ListMapIterRep<key,value>((ListMap<key,value> *)(this->Container));
    return ret;
  }

  //
  // These constructors allow a ListMapIter to be constructed from a
  // ListMap.
  //
  //+grp
  ListMapIterRep(ListMap<key,value> *st) : MapIterRep<key,value>(st), listp(st->list()) {}
  ListMapIterRep(ListMap<key,value> &st) : MapIterRep<key,value>(st), listp(st.list()) {}
  //-grp

  enum {ListMapIterRepVersion = 1};

};

} //#End casa namespace

#ifndef AIPS_NO_TEMPLATE_SRC
#include <casa/Containers/ListMap.cc>
#endif //# AIPS_NO_TEMPLATE_SRC
#endif
