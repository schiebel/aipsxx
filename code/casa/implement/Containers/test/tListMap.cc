//# tListMap.cc: This program tests the ListMap class
//# Copyright (C) 1993,1994,1995,1999,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: tListMap.cc,v 19.4 2004/11/30 17:50:16 ddebonis Exp $

//# Includes

#include <casa/Containers/ListMap.h>
#include <casa/BasicSL/String.h>
#include <casa/Containers/OrdPairIO.h>
#include <casa/Containers/MapIO.h>
#include <casa/iostream.h>

#include <casa/namespace.h>
//template<class key,class value> void print(const ListMap<key,value> &);

main() {
  ListMap<int,int> map(-1);
  ListMap<String,OrderedPair<String,uInt> > smap(OrderedPair<String,uInt>("dummy",0));
  ListMap<String,int> amap(-1,ListMap<String,int>::Append);

  map(2) = 90;
  map(15) = 79;
  map(27) = 104;
  map(21) = map(27);

  map.setOrder(ListMap<int,int>::Append);
  map(81) = 4;
  map(21) = 109;
  map(29) = map(21);
  map(28) = map(29);
  map(4) = 9;
  map.setOrder(ListMap<int,int>::Prepend);
  map(30) = 4;
  map(6) = 92;

  cout << map << endl;

//print(map);

  smap("fred") = OrderedPair<String,uInt>("wilma",30);
  smap("barney") = OrderedPair<String,uInt>("betty",28);
  smap("homer")  = OrderedPair<String,uInt>("marge",40);
  smap("bambam") = OrderedPair<String,uInt>("pebbles",8);

  cout << smap << endl;

//print(smap);

  smap("barney") = OrderedPair<String,uInt>("madona",25);
  smap("fred") = OrderedPair<String,uInt>("anne",12);
 
  cout << smap << endl;

//print(smap);
//print(smap);

  cout << smap("homer").x() << "," << smap("homer").y() << endl;
  cout << smap("fred").x() << "," << smap("fred").y() << endl;
  cout << smap("barney").x() << "," << smap("barney").y() << endl;

  amap("a") = 97;
  amap("b") = 98;
  amap("c") = 99;
  amap("d") = 100;
  amap("e") = 101;
  amap("f") = 102;

  cout << amap << endl;

//print(amap);

  return(0);

}

//#
//# Moved here to work around YASSB (yet another stupid sun bug)
//#
// template<class key,class value> void print(const ListMap<key,value> &xx){
//   ConstListMapIter<key,value> x(xx);
//   x.toStart();
//   while (!x.atEnd()) {
//     cout << "(" << x.getKey() << "," << x.getVal() << ")" << " ";
//     x++;
//   }
//   cout << endl;
// }
