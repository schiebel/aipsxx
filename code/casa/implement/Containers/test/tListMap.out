(6,92)(30,4)(21,109)(27,104)(15,79)(2,90)(81,4)(29,109)(28,109)(4,9)
(bambam,(pebbles,8))(homer,(marge,40))(barney,(betty,28))(fred,(wilma,30))
(bambam,(pebbles,8))(homer,(marge,40))(barney,(madona,25))(fred,(anne,12))
marge,40
anne,12
madona,25
(a,97)(b,98)(c,99)(d,100)(e,101)(f,102)
